### PR TITLE
Kubevirt csi driver integration

### PR DESCRIPTION
**Pull request by mfranczy**
**What this PR does / why we need it**:
We need it to provision storage for user clusters created on top of KubeVirt

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Support for KubeVirt CSI driver
```

### CODE REVIEW

**Comment by mfranczy** [comment]
/hold

I have to pass csi driver kubeconfig in a bit different way

File: pkg/provider/cloud/kubevirt/provider.go [hunk 1]
```diff
@@ -44,30 +43,37 @@ func (k *kubevirt) DefaultCloudSpec(spec *v1.CloudSpec) error {
...
 	}
 
-	config, err := base64.StdEncoding.DecodeString(kubeconfig)
+	return nil
+}
+
+func (k *kubevirt) InitializeCloudProvider(cluster *v1.Cluster, update provider.ClusterUpdater) (*v1.Cluster, error) {
+	ctx := context.Background()
+	// soon the namespace will be changed to the cluster namespace
+	namespace := "default"
```
Comment (hdurand0710) [inline pkg/provider/cloud/kubevirt/provider.go]: <3

File: pkg/addon/template.go [hunk 2]
```diff
@@ -229,6 +237,16 @@ type MLASettings struct {
 	LoggingEnabled bool
 }
 
+type CloudProviderData struct {
+	// KubeVirt specific data
+	KubeVirtData KubeVirtData
+}
+
+type KubeVirtData struct {
+	// StorageProvisionerKubeconfig represents a kubeconfig for user cluster to handle CSI driver requests
```
Comment (hdurand0710) [inline pkg/addon/template.go]: user cluster is the Kubevirt infra cluster when not deployed via Addons right ?

File: pkg/provider/cloud/kubevirt/client.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+type Clientset struct {
+	RestConfig *restclient.Config
+	// KubeVirt clientset
+	KubeVirt kv.Interface
+	// Kubernetes clientset
+	Kubernetes kubernetes.Interface
+}
+
+func NewClientset(cluster *v1.Cluster, secretKeyGetter provider.SecretKeySelectorValueFunc) (*Clientset, error) {
```
Comment (hdurand0710) [inline pkg/provider/cloud/kubevirt/client.go]: As you said in your previous comments, I would need a slightly different signature to reuse it for the VmiPreset and StorageClass.

**Comment by hdurand0710** [comment]
Looks good to me, just waiting for the ClientSet.

**Review by hdurand0710 — APPROVED** [review]

File: pkg/provider/cloud/kubevirt/client.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	// KubeVirt clientset
+	KubeVirt kv.Interface
+	// Kubernetes clientset
+	Kubernetes kubernetes.Interface
+}
+
+func NewClientset(kubeconfig string) (*Clientset, error) {
+	config, err := base64.StdEncoding.DecodeString(kubeconfig)
+	if err != nil {
+		// if the decoding failed, the kubeconfig is sent already decoded without the need of decoding it,
```
Comment (moadqassem) [inline pkg/provider/cloud/kubevirt/client.go:30]: So the problem with that is, in the past we had some issues passing and parsing the Kubeconfig as a json or yaml from UI to API and vice versa. Thus we demanded that this should be base64 encoded so we avoid all the issues that comes a long parsing data as plaintext

File: pkg/provider/cloud/kubevirt/storage.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func EnsureInfraCSIAccess(ctx context.Context, kubeconfig string, clusterName string) ([]byte, error) {
+	// soon we are going to change the default namespace to cluster namespace
```
Comment (moadqassem) [inline pkg/provider/cloud/kubevirt/storage.go]: What is the issue of doing it now? I mean the name space of the cluster is provided already to the function ;) 
`clusterName string` => `cluster-[clusterName]`

File: pkg/provider/cloud/kubevirt/storage.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	namespace := "default"
+
+	cli, err := NewClientset(kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	sa := &corev1.ServiceAccount{
```
Comment (moadqassem) [inline pkg/provider/cloud/kubevirt/storage.go]: Do we know already what kinda resources the client would access? I mean giving access to all resources might impose a security issues.

File: pkg/provider/cloud/kubevirt/storage.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kubevirt-csi",
+			Namespace: namespace,
+		},
+	}
+	_, err = cli.Kubernetes.CoreV1().ServiceAccounts(namespace).Create(ctx, sa, metav1.CreateOptions{})
```
Comment (moadqassem) [inline pkg/provider/cloud/kubevirt/storage.go]: Why not using the reconciling functions for creating those resources? Same goes for all the resources that you are creating. https://github.com/kubermatic/kubermatic/blob/master/pkg/resources/reconciling/zz_generated_reconcile.go#L177-L178

File: pkg/provider/cloud/kubevirt/storage.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+				Verbs:     []string{"list"},
+			},
+			{
+				APIGroups: []string{"subresources.kubevirt.io"},
+				Resources: []string{"virtualmachineinstances/addvolume", "virtualmachineinstances/removevolume"},
+				Verbs:     []string{"update"},
+			},
+		},
+	}
+	_, err = cli.Kubernetes.RbacV1().Roles(namespace).Create(ctx, role, metav1.CreateOptions{})
```
Comment (moadqassem) [inline pkg/provider/cloud/kubevirt/storage.go]: Why not using this reconciling functions instead accessing the api directly: 
https://github.com/kubermatic/kubermatic/blob/master/pkg/resources/reconciling/zz_generated_reconcile.go#L476-L477

File: pkg/provider/cloud/kubevirt/storage.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "Role",
+			Name:     "kubevirt-csi",
+		},
+	}
+
+	_, err = cli.Kubernetes.RbacV1().RoleBindings(namespace).Create(ctx, roleBinding, metav1.CreateOptions{})
```
Comment (moadqassem) [inline pkg/provider/cloud/kubevirt/storage.go]: Same goes here, you can use this function type instead https://github.com/kubermatic/kubermatic/blob/master/pkg/resources/reconciling/zz_generated_reconcile.go#L513-L514

**Review by moadqassem — CHANGES_REQUESTED** [review]

File: pkg/provider/cloud/kubevirt/storage.go (hunk 5 above)
Comment (mfranczy) [reply pkg/provider/cloud/kubevirt/storage.go]: I would have to change this for KubeVirt CCM, would be better to do it in the future PRs.

File: pkg/provider/cloud/kubevirt/storage.go (hunk 6 above)
Comment (mfranczy) [reply pkg/provider/cloud/kubevirt/storage.go:39]: Good idea, was not aware of those.

File: pkg/provider/cloud/kubevirt/client.go (hunk 3 above)
Comment (mfranczy) [reply pkg/provider/cloud/kubevirt/client.go]: We should merge it in this way and adjust in your PR.

File: pkg/addon/template.go (hunk 2 above)
Comment (mfranczy) [reply pkg/addon/template.go]: User cluster is a new k8s cluster created on top of KubeVirt

File: pkg/provider/cloud/kubevirt/storage.go (hunk 6 above)
Comment (mfranczy) [reply pkg/provider/cloud/kubevirt/storage.go]: I am not giving an acess to all resources here.

File: pkg/provider/cloud/kubevirt/storage.go (hunk 7 above)
Comment (mfranczy) [reply pkg/provider/cloud/kubevirt/storage.go]: I was not aware of those, good idea.

File: pkg/provider/kubernetes/credentials.go [hunk 10]
```diff
@@ -360,15 +361,25 @@ func createOrUpdatePacketSecret(ctx context.Context, seedClient ctrlruntimeclien
...
 
+	// ensure that CSI driver on user cluster will have an access to KubeVirt cluster
+	r, err := kubevirt.NewReconciler(spec.Kubeconfig, cluster.Name)
+	if err != nil {
+		return err
+	}
+	csiKubeconfig, err := r.ReconcileCSIAccess(ctx)
+	if err != nil {
+		return err
+	}
```
Comment (mfranczy) [inline pkg/provider/kubernetes/credentials.go:377]: With the current design I am concerned about that part. There is only one request to create a cluster and potentially we might hit a race condition inside `ReconcileCSIAccess`. It is a small risk, although still possible. As far as I can see KubeVirt is the first provider that automatically generate secrets for user cluster components.

Refactoring of the whole mechanism would be much bigger than scope of the issue. I would like to talk about this code with @zreigz on Monday. If we see that this problem occurs often then we will have to do refactoring or implement a better mechanism to generate KubeVirt secrets.

Maybe we should implement a proper reconciler for KubeVirt cloud provider and from there inject a kubeconfig for CSI driver, although that opens a new way to interact with user clusters (we skip addon mechanism)

File: pkg/provider/cloud/kubevirt/storage.go (hunk 5 above)
Comment (mfranczy) [reply pkg/provider/cloud/kubevirt/storage.go]: the same applies to the UI

File: pkg/provider/kubernetes/credentials.go (hunk 10 above)
Comment (mfranczy) [reply pkg/provider/kubernetes/credentials.go:377]: as discussed with @zreigz it's ok for now to leave it like this.

**Pull request merged**